cmd/kubeadm/app/phases/controlplane/components.go

```diff
@@ -0,0 +1,183 @@
+	"time"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	kubeletconfig "k8s.io/kubelet/config/v1beta1"
+
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util/staticpod"
+)
+
+// ControlPlaneComponents contains all components in control plane
+var ControlPlaneComponents = []string{
+	kubeadmconstants.KubeAPIServer,
+	kubeadmconstants.KubeControllerManager,
+	kubeadmconstants.KubeScheduler,
+}
+
+type component struct {
+	name    string
+	labels  map[string]string
+	touched bool
+}
+
+// WaitForControlPlaneComponents wait for control plane component to be ready by check pod status returned by kubelet
+func WaitForControlPlaneComponents(componentNames []string, timeout time.Duration, manifestDir, kubeletDir, certificatesDir string) error {
+	certFile := filepath.Join(certificatesDir, kubeadmconstants.APIServerKubeletClientCertName)
+	keyFile := filepath.Join(certificatesDir, kubeadmconstants.APIServerKubeletClientKeyName)
+
+	client, err := rest.HTTPClientFor(&rest.Config{
+		TLSClientConfig: rest.TLSClientConfig{
+			CertFile: certFile,
+			KeyFile:  keyFile,
+			Insecure: true,
+		},
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to create kubelet client")
+	}
+
+	kubeletEndpoint, err := getKubeletEndpoint(filepath.Join(kubeletDir, kubeadmconstants.KubeletConfigurationFileName))
+	if err != nil {
+		return errors.Wrap(err, "failed to get kubelet endpoint")
+	}
+
+	components := make([]*component, len(componentNames))
+	for i, name := range componentNames {
+		labels, err := getComponentLabels(name, manifestDir)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get pod labels of %s component", name)
+		}
+
+		components[i] = &component{name, labels, false}
+	}
+
+	return wait.PollUntilContextTimeout(context.Background(), 5*time.Second, timeout, false, func(ctx context.Context) (bool, error) {
+		klog.V(1).Infoln("[control-plane] polling status of control plane components...")
+
+		resp, err := client.Get(kubeletEndpoint)
+		if err != nil {
+			fmt.Printf("[kubelet client] Error getting pods [%v]\n", err)
+			return false, nil
+		}
+
+		defer resp.Body.Close()
+
+		data, err := io.ReadAll(resp.Body)
+		if err != nil {
+			fmt.Printf("[kubelet client] Error reading pods from response body [%v]\n", err)
+			return false, nil
+		}
+
+		pods := &v1.PodList{}
+		if err := json.Unmarshal(data, pods); err != nil {
+			fmt.Printf("[kubelet client] Error parsing pods from response body: %q\n", data)
+			return false, nil
+		}
+
+		for _, comp := range components {
+			labels := comp.labels
+		match_pod:
+			for _, pod := range pods.Items {
+				podLabels := pod.ObjectMeta.Labels
+				for key, value := range labels {
+					if podLabels[key] != value {
+						continue match_pod
+					}
+				}
+
+				comp.touched = true
```

Comment:
on a quick look it's not obvious why we the code is tracking labels and touched state. can you elaborate why status.Ready is not enough?

Reply:
> on a quick look it's not obvious why we the code is tracking labels and touched state. can you elaborate why status.Ready is not enough?

`component.touched: bool` is dealing with the case: when pods info returned from kubelet cannot be found with this `component.labels`.
I will add some comments on `component` struct

Reply:
ok, i will check the comments later.
currently, i still don't get why we need to the labels and touched flags.